src/ordering.f90

```diff
@@ -0,0 +1,98 @@
+module m_ordering
+
+   use m_common, only: dp, dir_X, dir_Y, dir_Z, &
+                       RDR_X2Y, RDR_X2Z, RDR_Y2X, RDR_Y2Z, RDR_Z2X, RDR_Z2Y
+
+   implicit none
+contains
+   !!
+   !! "Application storage" stores spatial data with a directionality for better cache locality
+   !!  This set of functions converts indices from this application storage (_dir) to cartesian indices (_ijk)
+   !! 
+
+   ! Get cartesian index from application storage directional one
+   pure subroutine get_index_ijk(i, j, k, dir_i, dir_j, dir_k, dir, SZ, nx_loc, ny_loc, nz_loc)
+      integer, intent(out) :: i, j, k                   ! cartesian indices
+      integer, intent(in) :: dir_i, dir_j, dir_k        ! application storage indices
+      integer, intent(in) :: dir                        ! direction of the applicatino storage indices
+      integer, intent(in) :: SZ, nx_loc, ny_loc, nz_loc ! dimensions of the block
+      
+      select case (dir)
+         case (dir_X)
+            i = dir_j
+            j = mod(dir_k - 1, ny_loc/SZ)*SZ + dir_i 
+            k = ceiling(real(dir_k)/(ny_loc/SZ))
```

Comment:
I was thinking if there is any way to avoid switching to reals. Because both the inputs and outputs are integers, it can be safer to keep everything as integers in the intermediate operations as well. For example maybe this can work here `1 + (dir_k - 1)/(ny_loc/SZ)`

Reply:
Good point, I will look into it